node/src/components/linear_chain_sync/operations.rs

```diff
@@ -484,19 +521,162 @@ pub(crate) async fn run_fast_sync_task(
+///  3. Fetches all deploys and finality signatures.
+///  4. Stops after fetching a block with the current version and its global state.
+///
+/// Returns the block header with our current version and the last trusted key block information for
+/// validation.
+async fn archival_sync(
+    effect_builder: EffectBuilder<JoinerEvent>,
+    trusted_block_header: BlockHeader,
+    chainspec: &Chainspec,
+) -> Result<(KeyBlockInfo, BlockHeader), LinearChainSyncError> {
+    // Get the trusted block info. This will fail if we are trying to join with a trusted hash in
+    // era 0.
+    let mut trusted_key_block_info =
+        get_trusted_key_block_info(effect_builder, chainspec, &trusted_block_header).await?;
+
+    let trusted_block =
+        *fetch_and_store_block_by_hash(effect_builder, trusted_block_header.hash()).await?;
+
+    // Sync to genesis
+    let mut walkback_block = trusted_block.clone();
+    loop {
+        sync_deploys_and_transfers_and_state(effect_builder, &walkback_block).await?;
+        if walkback_block.height() == 0 {
+            break;
+        } else {
+            walkback_block = *fetch_and_store_block_by_hash(
+                effect_builder,
+                *walkback_block.header().parent_hash(),
+            )
+            .await?;
+        }
+    }
+
+    // Sync forward until we are at the current version.
+    let mut most_recent_block = trusted_block;
+    while most_recent_block.header().protocol_version() < chainspec.protocol_config.version {
+        let maybe_fetched_block_with_metadata = fetch_and_store_block_by_height(
+            effect_builder,
+            most_recent_block.header().height() + 1,
+            &trusted_key_block_info,
+            chainspec.highway_config.finality_threshold_fraction,
+        )
+        .await?;
+        most_recent_block = match maybe_fetched_block_with_metadata {
+            Some(block_with_metadata) => block_with_metadata.block,
+            None => {
+                tokio::time::sleep(SLEEP_DURATION_SO_WE_DONT_SPAM).await;
+                continue;
+            }
+        };
+        sync_deploys_and_transfers_and_state(effect_builder, &most_recent_block).await?;
+        if let Some(key_block_info) =
+            KeyBlockInfo::maybe_from_block_header(most_recent_block.header())
+        {
+            trusted_key_block_info = key_block_info;
+        }
+    }
+
+    Ok((trusted_key_block_info, most_recent_block.take_header()))
+}
+
+/// Runs the fast synchronization task.
+pub(crate) async fn run_fast_sync_task(
+    effect_builder: EffectBuilder<JoinerEvent>,
+    trusted_hash: BlockHash,
+    chainspec: Arc<Chainspec>,
+    node_config: NodeConfig,
+) -> Result<BlockHeader, LinearChainSyncError> {
+    // Fetch the trusted header
+    let trusted_block_header = fetch_and_store_block_header(effect_builder, trusted_hash).await?;
+
+    if let Some(trusted_hash_time_to_expiration) = node_config.trusted_hash_time_to_expiration {
```

Comment:
This should be computed from the unbonding delay and auction delay, otherwise the configuration could go out of sync with the actual delay for withdrawing stake.

Reply:
Should be fixed in https://github.com/casper-network/casper-node/commit/fe6cd1be96a832d660077068a208a5eaa5f1135e